File: ios/Detonator/Sources/Detonator/Module/StyleSheetModule/Style.swift
public enum StyleColor: Decodable {
    case string(String)
    case array([Float])
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let string = try? container.decode(String.self) {
            self = .string(string)
        } else if let array = try? container.decode([Float].self) {
            self = .array(array)
        } else {
            throw DecodingError.typeMismatch(StyleColor.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Expected String or [Float]"))
        }
    }
}

public enum StyleSize: Decodable {
    case float(Float)
    case string(String)
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let float = try? container.decode(Float.self) {
            self = .float(float)
        } else if let string = try? container.decode(String.self) {
            self = .string(string)
        } else {
            throw DecodingError.typeMismatch(StyleSize.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Expected Float or String"))
        }
    }
}

public struct StyleTransform: Decodable {
    var translateX: StyleSize?
    var translateY: StyleSize?
    var scale: StyleSize?
    var scaleX: StyleSize?
    var scaleY: StyleSize?
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        translateX = try container.decodeIfPresent(StyleSize.self, forKey: .translateX)
        translateY = try container.decodeIfPresent(StyleSize.self, forKey: .translateY)
        scale = try container.decodeIfPresent(StyleSize.self, forKey: .scale)
        scaleX = try container.decodeIfPresent(StyleSize.self, forKey: .scaleX)
        scaleY = try container.decodeIfPresent(StyleSize.self, forKey: .scaleY)
    }
    
    private enum CodingKeys: String, CodingKey {
        case translateX
        case translateY
        case scale
        case scaleX
        case scaleY
    }
}

public struct Style: Decodable {
    // flex
    
    var flex: Int?
    var flexDirection: String?
    var justifyContent: String?
    var alignItems: String?
    var alignSelf: String?
    var gap: Float?
    
    // bg
    
    var backgroundColor: StyleColor?
    
    // size
    
    var width: StyleSize?
    var height: StyleSize?
    var minWidth: StyleSize?
    var minHeight: StyleSize?
    var maxWidth: StyleSize?
    var maxHeight: StyleSize?
    
    // padding
    
    var padding: Float?
    var paddingHorizontal: Float?
    var paddingVertical: Float?
    var paddingTop: Float?
    var paddingLeft: Float?
    var paddingBottom: Float?
    var paddingRight: Float?
    
    // margin
    
    var margin: Float?
    var marginHorizontal: Float?
    var marginVertical: Float?
    var marginTop: Float?
    var marginLeft: Float?
    var marginBottom: Float?
    var marginRight: Float?
    
    // text
    
    var fontSize: Float?
    var lineHeight: Float?
    var fontWeight: String?
    var color: StyleColor?
    var textAlign: String?
    var textDecoration: String?
    var textTransform: String?
    var textOverflow: String?
    var overflowWrap: String?
    var wordBreak: String?
    
    // position
    
    var position: String?
    var top: StyleSize?
    var left: StyleSize?
    var bottom: StyleSize?
    var right: StyleSize?
    var zIndex: Int?
    
    // misc
    
    var display: String?
    var pointerEvents: String?
    var objectFit: String?
    var overflow: String?
    var opacity: Float?
    var aspectRatio: Float?
    
    // border radius
    
    var borderRadius: StyleSize?
    var borderRadiusTopLeft: StyleSize?
    var borderRadiusTopRight: StyleSize?
    var borderRadiusBottomLeft: StyleSize?
    var borderRadiusBottomRight: StyleSize?
    
    // border
    
    var borderWidth: Float?
    var borderColor: StyleColor?
    
    var borderTopWidth: Float?
    var borderTopColor: StyleColor?
    
    var borderLeftWidth: Float?
    var borderLeftColor: StyleColor?
    
    var borderBottomWidth: Float?
    var borderBottomColor: StyleColor?
    
    var borderRightWidth: Float?
    var borderRightColor: StyleColor?
    
    // transform
    
    var transform: StyleTransform?
}

File: android/Detonator/src/main/java/com/iconshot/detonator/module/stylesheet/Style.java
package com.iconshot.detonator.module.stylesheet;

public class Style {
    // flex

    public Integer flex;
    public String flexDirection;
    public String justifyContent;
    public String alignItems;
    public String alignSelf;
    public Float gap;

    // bg

    public Object backgroundColor;

    // size

    public Object width;
    public Object height;
    public Object minWidth;
    public Object minHeight;
    public Object maxWidth;
    public Object maxHeight;

    // padding

    public Float padding;
    public Float paddingHorizontal;
    public Float paddingVertical;
    public Float paddingTop;
    public Float paddingLeft;
    public Float paddingBottom;
    public Float paddingRight;

    // margin

    public Float margin;
    public Float marginHorizontal;
    public Float marginVertical;
    public Float marginTop;
    public Float marginLeft;
    public Float marginBottom;
    public Float marginRight;

    // text

    public Float fontSize;
    public Float lineHeight;
    public String fontWeight;
    public Object color;
    public String textAlign;
    public String textDecoration;
    public String textTransform;
    public String textOverflow;
    public String overflowWrap;
    public String wordBreak;

    // position

    public String position;
    public Object top;
    public Object left;
    public Object bottom;
    public Object right;
    public Integer zIndex;

    // misc

    public String display;
    public String pointerEvents;
    public String objectFit;
    public String overflow;
    public Float opacity;
    public Float aspectRatio;

    // border radius

    public Object borderRadius;
    public Object borderRadiusTopLeft;
    public Object borderRadiusTopRight;
    public Object borderRadiusBottomLeft;
    public Object borderRadiusBottomRight;

    // border

    public Float borderWidth;
    public Object borderColor;

    public Float borderTopWidth;
    public Object borderTopColor;

    public Float borderLeftWidth;
    public Object borderLeftColor;

    public Float borderBottomWidth;
    public Object borderBottomColor;

    public Float borderRightWidth;
    public Object borderRightColor;

    // transform

    public StyleTransform transform;

    public static class StyleTransform {
        public Object translateX;
        public Object translateY;
        public Object scale;
        public Object scaleX;
        public Object scaleY;
    }
}